TYPO3.Flow/Classes/TYPO3/Flow/Package/Package.php

```diff
@@ -425,97 +417,159 @@ public function getDocumentationPath()
+        return Utility::getComposerManifest($this->packagePath, $key);
     }
 
     /**
-     * Builds and returns an array of class names => file names of all
-     * *.php files in the package's Classes directory and its sub-
-     * directories.
+     * Get the installed package version (from composer)
      *
-     * @param string $classesPath Base path acting as the parent directory for potential class files
-     * @param string $extraNamespaceSegment A PHP class namespace segment which should be inserted like so: \TYPO3\PackageKey\{namespacePrefix\}PathSegment\PathSegment\Filename
-     * @param string $subDirectory Used internally
-     * @param integer $recursionLevel Used internally
-     * @return array
-     * @throws \TYPO3\Flow\Package\Exception if recursion into directories was too deep or another error occurred
+     * @return string
+     * @api
+     * TODO: Should be added to the interface in the next major Flow version (4.0)
      */
-    protected function buildArrayOfClassFiles($classesPath, $extraNamespaceSegment = '', $subDirectory = '', $recursionLevel = 0)
+    public function getInstalledVersion()
     {
-        $classFiles = array();
-        $currentPath = $classesPath . $subDirectory;
-        $currentRelativePath = substr($currentPath, strlen($this->packagePath));
-        $namespacePrefix = '';
+        return PackageManager::getPackageVersion($this->composerName);
+    }
 
-        if ($this->getAutoloadType() === self::AUTOLOADER_TYPE_PSR4) {
-            $namespacePrefix = $this->getNamespace();
+    /**
+     * @param string $autoloadType
+     * @param string $autoloadNamespace
+     * @param string $autoloadPath
+     * @return string
+     */
+    protected function normalizeAutoloadPath($autoloadType, $autoloadNamespace, $autoloadPath)
+    {
+        $normalizedAutoloadPath = $autoloadPath;
+        if ($autoloadType === ClassLoader::MAPPING_TYPE_PSR0) {
+            $normalizedAutoloadPath = Files::concatenatePaths([
+                    $autoloadPath,
+                    str_replace('\\', '/', $autoloadNamespace)
+                ]) . '/';
         }
 
-        if (!is_dir($currentPath)) {
-            return array();
-        }
-        if ($recursionLevel > 100) {
-            throw new Exception('Recursion too deep while collecting class files.', 1166635495);
+        return $normalizedAutoloadPath;
+    }
+
+    /**
+     * @param string $baseAutoloadPath
+     * @param string $autoloadNamespace
+     * @return \Generator
+     */
+    protected function getClassesInNormalizedAutoloadPath($baseAutoloadPath, $autoloadNamespace)
+    {
+        $autoloadNamespace = trim($autoloadNamespace, '\\') . '\\';
+        $directories = [''];
+        while ($directories !== []) {
+            $currentRelativeDirectory = array_pop($directories);
+            $currentAbsoluteDirectory = $baseAutoloadPath . $currentRelativeDirectory;
+            if ($handle = opendir($currentAbsoluteDirectory)) {
```

Comment:
a specific reason why you went back to opendir/readdir/closedir over DirectoryIterator?